src/main/java/org/opensearch/knn/memoryoptsearch/faiss/FaissMemoryOptimizedSearcher.java

```diff
@@ -16,32 +23,91 @@
  * This searcher directly reads FAISS index file via the provided {@link IndexInput} then perform vector search on it.
  */
 public class FaissMemoryOptimizedSearcher implements VectorSearcher {
+    private static final FlatVectorsScorer VECTOR_SCORER = FlatVectorScorerUtil.getLucene99FlatVectorsScorer();
+
     private final IndexInput indexInput;
     private FaissIndex faissIndex;
+    private FaissHnswGraph faissHnswGraph;
 
     public FaissMemoryOptimizedSearcher(IndexInput indexInput) throws IOException {
         this.indexInput = indexInput;
-        try {
-            this.faissIndex = FaissIndex.load(indexInput);
-        } catch (UnsupportedFaissIndexException e) {
-            // TODO(KDY) : Remove this in part-7 (Complete Faiss Loading Part at Codec Level)
+        this.faissIndex = FaissIndex.load(indexInput);
+        final FaissHNSW hnsw = extractFaissHnsw(faissIndex);
+        this.faissHnswGraph = new FaissHnswGraph(hnsw, indexInput);
+    }
+
+    private static FaissHNSW extractFaissHnsw(final FaissIndex faissIndex) {
+        if (faissIndex instanceof FaissIdMapIndex idMapIndex) {
+            return idMapIndex.getNestedIndex().getHnsw();
         }
+
+        throw new IllegalArgumentException("Faiss index [" + faissIndex.getIndexType() + "] does not have HNSW as an index.");
     }
 
     @Override
     public void search(float[] target, KnnCollector knnCollector, Bits acceptDocs) throws IOException {
-        // TODO(KDY) : This will be covered in part-7 (Complete Faiss Loading Part at Codec Level)
-        throw new UnsupportedOperationException("Not implemented yet");
+        search(
+            VectorEncoding.FLOAT32,
+            () -> VECTOR_SCORER.getRandomVectorScorer(
+                faissIndex.getVectorSimilarityFunction().getVectorSimilarityFunction(),
+                faissIndex.getFloatValues(indexInput),
```

Comment:
Are we opening up the flatVectorReaders as well already? Can we avoid one of the two?